experiments/tagging/embedding.py

```diff
@@ -0,0 +1,271 @@
+        scalars = torch.cat((scalars, pt), dim=-1)
+
+    # embed fourmomenta into multivectors
+    multivectors = embed_vector(fourmomenta)
+    multivectors = multivectors.unsqueeze(-2)
+
+    # beam reference
+    spurions = get_spurion(
+        cfg_data.beam_reference,
+        cfg_data.add_time_reference,
+        cfg_data.two_beams,
+        fourmomenta.device,
+        fourmomenta.dtype,
+    )
+    n_spurions = spurions.shape[0]
+    if cfg_data.beam_token:
+        # prepend spurions to the token list (within each block)
+        spurion_idxs = torch.cat(
+            [
+                torch.arange(
+                    ptr_start + i * n_spurions,
+                    ptr_start + (i + 1) * n_spurions,
+                )
+                for i, ptr_start in enumerate(ptr[:-1])
+            ]
+        )
+        insert_spurion = torch.zeros(
+            multivectors.shape[0] + n_spurions * batchsize,
+            dtype=torch.bool,
+            device=multivectors.device,
+        )
+        insert_spurion[spurion_idxs] = True
+        multivectors_buffer = multivectors.clone()
+        multivectors = torch.empty(
+            insert_spurion.shape[0],
+            *multivectors.shape[1:],
+            dtype=multivectors.dtype,
+            device=multivectors.device,
+        )
+        multivectors[~insert_spurion] = multivectors_buffer
+        multivectors[insert_spurion] = spurions.repeat(batchsize, 1).unsqueeze(-2)
+        scalars_buffer = scalars.clone()
+        scalars = torch.zeros(
+            multivectors.shape[0],
+            scalars.shape[1],
+            dtype=scalars.dtype,
+            device=scalars.device,
+        )
+        scalars[~insert_spurion] = scalars_buffer
+        ptr[1:] = ptr[1:] + (arange + 1) * n_spurions
+    else:
+        # append spurion to multivector channels
+        spurions = spurions.unsqueeze(0).repeat(multivectors.shape[0], 1, 1)
+        multivectors = torch.cat((multivectors, spurions), dim=-2)
+
+    # global tokens
+    if cfg_data.include_global_token:
+        # prepend global tokens to the token list
+        num_global_tokens = cfg_data.num_global_tokens
+        global_idxs = torch.cat(
+            [
+                torch.arange(
+                    ptr_start + i * num_global_tokens,
+                    ptr_start + (i + 1) * num_global_tokens,
+                )
+                for i, ptr_start in enumerate(ptr[:-1])
+            ]
+        )
+        is_global = torch.zeros(
+            multivectors.shape[0] + batchsize * num_global_tokens,
+            dtype=torch.bool,
+            device=multivectors.device,
+        )
+        is_global[global_idxs] = True
+        multivectors_buffer = multivectors.clone()
+        multivectors = torch.zeros(
+            is_global.shape[0],
+            *multivectors.shape[1:],
+            dtype=multivectors.dtype,
+            device=multivectors.device,
+        )
+        multivectors[~is_global] = multivectors_buffer
```

Comment:
Before this line, we should add the following lines to include a 1 in the scalar component of the multivector global token:

```suggestion
        global_mv = torch.zeros((1, multivectors.shape[-1]), device=multivectors.device)
        global_mv[:,0] = 1
        multivectors[is_global] = global_mv.repeat(batchsize * num_global_tokens, 1).unsqueeze(-2)
        multivectors[~is_global] = multivectors_buffer

Reply:
This information is already there in `scalars` (see the `token_idx` a few lines below), so no need to include it a second time. 

Your suggestion only works if we have 1 global token (`data.score_token=false`), in case `data.score_token=true`/10 global tokens you would need 10 multivector channels to have enough scalar channels where you can embed this information (similar to how we do it with the `token_idx` right now in `scalar` channels).